### ISSUE TITLE
Map the real Needle learning interface and runtime requirements

### ISSUE THREAD

**Issue by TeaShaman-cyber** [TeaShaman-cyber/theseus-needle-lab#1]
Determine the real executable Needle learning interface before any training CI is written. Record entry points, dependency versions, accepted data formats, CPU/GPU requirements, expected duration, outputs/artifacts, random-seed and reproducibility controls, and measured suitability of GitHub-hosted runners. Do not assume hosted runners are sufficient before measurement.

**Comment by TeaShaman-cyber** [comment]
## Interface investigation checkpoint — 2026-08-30

**Upstream source:** `cactus-compute/needle` at commit `ee221ce7c13579d9809209b979a9b7a50936614c` (package `cactus-needle` 2.0.8).

### Verified executable interface

Needle 2 fine-tuning is not pretraining. The supported user path is:

`JSONL -> needle finetune -> LoRA adapter .pkl -> needle build -> tuned .cact`

Training dependencies are the `train` extra: numpy, jax/jaxlib, flax, optax, sentencepiece. CPU Linux x86_64 is a supported JAX backend. NVIDIA and Apple accelerator extras are optional.

### Data contract

Each JSONL line contains `query`, `tools`, `answers`, and optional `reasoning` / `system`. Off-topic examples use `answers: []`. Default max length is 1024; upstream docs explicitly note that shorter examples reduce CPU cost substantially.

### Network / secret boundary

A first fine-tuning run does **not** require OpenRouter or any secret if `--generate` is not used. Missing base weights and tokenizer are fetched from the public `Cactus-Compute/needle2` Hugging Face repository. OpenRouter is only used by `generate-data` / `--generate`.

The current raw Needle 2 checkpoint is approximately 90.4 MB. The deployed `.cact` is approximately 13.7 MB.

### Reproducibility controls observed in source

- LoRA initialization uses JAX PRNG key 0.
- Validation split uses NumPy RNG seed 0.
- Training epoch shuffling currently uses NumPy global RNG without an explicit CLI seed, so exact training reproducibility is **not yet guaranteed** by the public CLI.
- Adapter output records rank, scale, base path and LoRA arrays.
- Tuned weights disable the base confidence head (`confidence=None`), because fine-tuning does not update that head.

### GitHub-hosted runner hypothesis

Standard public `ubuntu-latest` currently provides 4 CPU / 16 GB RAM / 14 GB SSD and is free for public repositories. GitHub-hosted jobs have a 6-hour execution limit. Memory/storage appear sufficient by static inspection; **CPU wall time remains UNKNOWN until measured**.

### Next verification

Run a bounded, secret-free GitHub Actions smoke experiment using a tiny hand-written JSONL dataset and the real Needle 2 checkpoint. Measure dependency install, checkpoint download, JAX compile, fine-tune wall time, peak memory where observable, adapter size, build wall time, `.cact` size, and SHA-256 outputs. This is a feasibility measurement, not a model-quality claim.

**Comment by TeaShaman-cyber** [comment]
## GitHub-hosted CPU smoke — VERIFIED 2026-08-30

The bounded secret-free experiment from this issue has now executed on a real standard public GitHub-hosted runner.

**Source branch:** `experiment/needle-cpu-smoke`  
**Commit:** `9a9b4bf1083a4fe2f8f732f779ede008ab1c3b75`  
**Workflow:** `Needle CPU smoke`  
**Run:** `33319821037` — **SUCCESS**

### Exact experiment

```text
cactus-needle[train] == 2.0.8
upstream reference = cactus-compute/needle@ee221ce7c13579d9809209b979a9b7a50936614c
examples = 12 hand-written public JSONL rows
max_len = 128
epochs = 1
batch = 2
LoRA rank = 4
--generate = 0
OpenRouter = not used
secrets = none
job timeout = 45 min
```

### Observed runner

```text
ubuntu image     ubuntu24 / 20260823.283.1
CPU              4
RAM MemTotal     16,372,432 KiB (~15.6 GiB)
Python           3.11.16
cactus-needle    2.0.8
JAX / jaxlib     0.10.2 / 0.10.2
Flax             0.12.8
Optax            0.2.8
```

### Measured stages

```text
dependency install      22 s
HF checkpoint download   5 s
finetune wall           23.59 s
finetune peak RSS        2,431,284 KiB (~2.32 GiB)
build wall               4.05 s
build peak RSS             663,752 KiB (~648 MiB)
```

The fine-tune executed 6 training steps plus the default 1-example validation holdout. Backend reported `cpu float32`.

### Artifacts and independent readback

```text
base checkpoint
  size   90,426,504 bytes
  sha256 4b0a972d163ffc7678fb3c36bace508114872e9d2ce9e10f225825752d3795bc

adapter.pkl
  size   1,991,592 bytes
  sha256 7005de88bbe7fa9cfaa3e7cab90fc344e2b9e5e45f187f5b90c75cf0c8f9e7fc

tuned.cact
  size   13,737,807 bytes
  sha256 3c0c684888c0d796e1b3a62326fbb1f3cc991f6ee5a0e596ac448df99edef10a
```

GitHub artifact `needle-cpu-smoke-33319821037` was downloaded after the run. Adapter and `.cact` size/SHA-256 were recomputed independently and matched `receipt.json` exactly.

### Disposition

**FACT:** a standard free public GitHub-hosted `ubuntu-latest` runner can complete this real Needle 2 `finetune -> build` smoke comfortably inside the bounded envelope.

**FACT:** memory is not the limiting resource for this tiny configuration; measured fine-tune peak RSS was ~2.32 GiB against ~15.6 GiB visible RAM.

**UNKNOWN:** scaling of CPU wall time and RSS with materially larger datasets, sequence lengths, epochs, batch sizes, and LoRA ranks remains unmeasured.

**Boundary:** this run is feasibility evidence only. It is not evidence that the 12-example adapter learned a useful policy or generalizes.

**Comment by TeaShaman-cyber** [comment]
## Implementation ledger — issue ↔ commits ↔ runs

This comment restores explicit traceability required by `docs/experiment-lifecycle.md` without rewriting already-pushed history.

| Commit | Purpose | Execution evidence | Disposition |
|---|---|---|---|
| `9a9b4bf1083a4fe2f8f732f779ede008ab1c3b75` | Add the first bounded real Needle CPU `finetune -> build` smoke, 12-example public dataset, receipt writer, pinned Actions, and contract tests | Actions run `33319821037` | **VERIFIED:** real CPU smoke completed; adapter and `.cact` hashes independently matched receipt |

Future commits for this issue line should include `refs #1` in the commit subject/body. `closes #1` is reserved for an explicit final disposition, not merely a green workflow.

**Comment by TeaShaman-cyber** [comment]
## Reproducibility finding — finetune epoch shuffle is not seed-addressable in CLI

Fresh source read of pinned upstream:

```text
cactus-compute/needle@ee221ce7c13579d9809209b979a9b7a50936614c
cactus-needle 2.0.8
```

shows an important training-interface detail.

### Deterministic pieces

```text
LoRA init:
  jax.random.PRNGKey(0)

validation split:
  np.random.default_rng(0).permutation(...)
```

### Non-deterministic piece

Each training epoch uses:

```python
order = np.random.permutation(count)
```

The `finetune` CLI has no `--seed` argument and the module does not set the global NumPy RNG seed before this loop.

`needle run` does expose `--seed`, but that is an inference/run control, not a finetune shuffle seed.

### Consequence

**FACT:** identical public `needle finetune` arguments do not currently establish a byte-reproducible adapter identity because minibatch ordering may differ across runs.

This does not invalidate the verified CPU-feasibility smoke, whose question was only whether real finetune→build can complete. It does matter for any subsequent quality comparison where training output identity or repeated-run variance is part of the claim.

### Experiment policy

A stronger-training quality probe must therefore do one of the following explicitly:

1. seed NumPy in a thin experiment wrapper before delegating to the pinned finetune implementation; or
2. run multiple replicas and treat shuffle variation as an experimental factor.

Do not silently call two adapters "the same training" based only on CLI arguments.

**Comment by TeaShaman-cyber** [comment]
## Training-contract audit follow-up — 2026-08-31

Read-only inspection of the pinned Needle 2.0.8 fine-tune path and the exact CPU-smoke configuration found three quality-experiment confounds that were not relevant to the original feasibility claim but do matter for causal interpretation of Issue #4.

### FACT — only five non-zero learning-rate parameter updates

The smoke used 12 rows with default `val_split=0.1`, leaving 11 training rows. With `batch_size=2` and `epochs=1`, Needle reports 6 optimizer steps and `warmup=1`.

Pinned Optax 0.2.8 applies the schedule at counter 0 on the first AdamW update. The warmup schedule starts at `init_value=0.0`, so the first parameter update has learning rate 0. Adam moments still update, but parameter motion occurs only on the remaining five updates.

### FACT — `max_len=128` truncates two training targets

The first independent reconstruction suggested a possible third boundary case, but the authoritative GitHub-hosted run with installed `cactus-needle==2.0.8` and its actual tokenizer resolved the exact set before training: rows 7 and 9 are truncated; row 12 fits exactly at 128 tokens with EOS retained.

- row 7 (`READY`): truncates inside the JSON tool call before closing quote/brackets;
- row 9 (`UNKNOWN`): truncates inside the JSON tool call before closing quote/brackets.

The deterministic validation split is row 10 (1-based), so both truncated rows are part of training.

Class-wise complete targets in the actual training set are therefore asymmetric:

```text
PROBE   4/4 complete
READY   3/4 complete
UNKNOWN 2/3 complete
```

This asymmetry is a plausible confound for the observed tendency of the tiny adapter to introduce extra `PROBE` outputs, but causality is not established.

### FACT — epoch shuffle is not seed-addressable through the public finetune CLI

This was already recorded on this Issue: `np.random.permutation(count)` uses global NumPy RNG without a finetune CLI seed. Repeating nominally identical training does not establish identical minibatch order.

### Consequence

The original CPU smoke remains valid feasibility evidence. These findings only constrain downstream model-quality experiments: a controlled comparison must fix shuffle seed and avoid target truncation before stronger-training claims are made.

**Comment by TeaShaman-cyber** [comment]
## Reproducibility follow-up — seeded CLI wrapper is not byte/deterministically behavioral across runs

Cross-run comparison found a new boundary while verifying #10.

Two nominally identical controls:
- #8 run `33428398968`, full-target / 1 epoch / seed 0 / max-len cap 1024
- #10 run `33432515778`, Arm A, same configuration

had identical package versions and identical printed training loss trajectory through all 6 steps and validation, but different adapter and W4 hashes.

Direct numeric comparison of the LoRA arrays shows real, very small float32 differences rather than pickle-only byte noise:

```text
max absolute element delta = 5.960464477539063e-08
changed numeric elements    = 205764 / 497664
```

After W4 export, behavior differed on:
- `1/12` training predictions (both wrong; accuracy unchanged at 1/12)
- `1/24` held-out predictions (both wrong; accuracy unchanged at 6/24)

**FACT:** setting the NumPy epoch-shuffle seed removes the known shuffle nondeterminism, but it does not establish byte-identical or prediction-identical training/export across separate GitHub-hosted runs.

**INFERENCE:** tiny floating-point execution differences can cross W4 quantization / decode decision boundaries. Future tiny-effect experiments should use same-run paired controls and treat one-prediction changes as an observed noise floor unless replicated.

**Comment by TeaShaman-cyber** [comment]
## Learning-interface invariant from #14

Verification of #14 exposed a prompt-contract property worth preserving in the learning-interface map:

**For Needle 2.0.8, serialized tool-schema order is behaviorally observable input. Semantic JSON equivalence is not sufficient for controlled experiments.**

A first #14 run accidentally used three schema serializations across the accepted #12 reference, written training JSONL, and evaluation. Base behavior shifted materially. After canonicalizing all corrected training/evaluation schemas to the exact #12 ordering (`name -> parameters -> description`), the base prediction vector reproduced #12 Arm D exactly (12/12).

Experiment contracts should therefore version/canonicalize the exact rendered tool schema or prompt bytes, not only the logical JSON Schema object.

**Comment by TeaShaman-cyber** [comment]
## Learning-interface invariant from #16

For Needle 2.0.8, **verbalizer surface form is part of the effective task contract, not cosmetic output formatting**.

On the same 12 inputs, same base model, same accepted framing and no training, changing only the enum representation materially changes both tool-call reachability and class distribution. Lowercase semantic labels outperform uppercase raw accuracy but introduce a strong PROBE prior; arbitrary one-token A/B/C labels reduce call reachability sharply.

Experiment receipts should therefore version exact enum strings/tokenization and report class prediction distributions, not only aggregate accuracy and schema validity.

**Comment by TeaShaman-cyber** [comment]
## Learning-interface invariant from #18 — label-token identity is behavioral state

#18 provides direct evidence that, for Needle 2.0.8 under grammar-constrained tool calling, the identity of label tokens is not a neutral encoding layer.

With fixed enum tokens `A/B/C` and all six semantic remappings, 10/12 inputs retained the same raw output across every mapping; token `B` was never emitted. Therefore experiment contracts must treat the exact verbalizer/token set and mapping as part of the behavioral interface, not as interchangeable class IDs.

This complements the existing schema-serialization invariant from #14: both serialized schema bytes and label-token identity are experimentally observable parts of the prompt contract.

**Comment by TeaShaman-cyber** [comment]
## Learning-interface currentness invariant from #26 Stage A

Stage A of #26 exposed a version-bound API difference before accepted inference evidence was collected.

Pinned experiment runtime `cactus-needle==2.0.8` does **not** export `needle.get_tokenizer()` from the package root. The exact 2.0.8 wheel exposes the tokenizer through `needle.model.tokenizer.get_tokenizer()` and `SANTokenizer.sp` wraps the underlying SentencePiece processor.

Current upstream documentation/source has since changed, so current-main API examples must not be assumed to describe pinned historical experiment packages.

**Invariant:** Needle experiment receipts and implementation tests must bind behavior to the exact package version/API surface under test; current upstream docs are not authority for a pinned older runtime.

The failed pre-evidence run is preserved in #26 audit history; accepted Stage A evidence was collected only after a regression test verified the 2.0.8 tokenizer path.
